crates/uv-resolver/src/resolver/environment.rs

```diff
@@ -0,0 +1,420 @@
+/// this module. The motivation for this design is to discourage repeated case
+/// analysis on this type, and instead try to encapsulate the case analysis via
+/// higher level routines on `ResolverEnvironment` itself. (This goal may prove
+/// intractable, so don't treat it like gospel.)
+#[derive(Clone, Debug)]
+enum Kind {
+    /// We're solving for one specific marker environment only.
+    ///
+    /// Generally, this is what's done for `uv pip`. For the project based
+    /// commands, like `uv lock`, we do universal resolution.
+    Specific {
+        /// The marker environment being resolved for.
+        ///
+        /// Any dependency specification that isn't satisfied by this marker
+        /// environment is ignored.
+        marker_env: ResolverMarkerEnvironment,
+    },
+    /// We're solving for all possible marker environments.
+    Universal {
+        /// The initial set of "fork preferences." These will come from the
+        /// lock file when available, or the list of supported environments
+        /// explicitly written into the `pyproject.toml`.
+        ///
+        /// Note that this may be empty, which means resolution should begin
+        /// with no forks. Or equivalently, a single fork whose marker
+        /// expression matches all marker environments.
+        initial_forks: Arc<[MarkerTree]>,
+        /// The markers associated with this resolver fork.
+        markers: MarkerTree,
+    },
+}
+
+impl ResolverEnvironment {
+    /// Create a resolver environment that is fixed to one and only one marker
+    /// environment.
+    ///
+    /// This enables `uv pip`-style resolutions. That is, the resolution
+    /// returned is only guaranteed to be installable for this specific marker
+    /// environment.
+    pub fn specific(marker_env: ResolverMarkerEnvironment) -> ResolverEnvironment {
+        let kind = Kind::Specific { marker_env };
+        ResolverEnvironment { kind }
+    }
+
+    /// Create a resolver environment for producing a multi-platform
+    /// resolution.
+    ///
+    /// The set of marker expressions given corresponds to an initial
+    /// seeded set of resolver branches. This might come from a lock file
+    /// corresponding to the set of forks produced by a previous resolution, or
+    /// it might come from a human crafted set of marker expressions.
+    ///
+    /// The "normal" case is that the initial forks are empty. When empty,
+    /// resolution will create forks as needed to deal with potentially
+    /// conflicting dependency specifications across distinct marker
+    /// environments.
+    ///
+    /// The order of the initial forks is significant. (It is unclear to AG
+    /// at time of writing what precisely is significant about it. Is this
+    /// just a matter of it being order dependent but otherwise no specific
+    /// guarantees are provided, or is precedence given to earlier forks over
+    /// later forks?)
```

Comment:
i see the order of forks similar to the order of dependencies: we don't guarantee any specific treatment, but reordering can help where our prioritization fails.

Reply:
Thanks! I've updated the wording here.